File: src/core/imageMetadata.ts
export interface ImageMetadata {
  width: number;
  height: number;
  format: "png" | "bmp" | "gif" | "jpeg" | "webp";
  animated: boolean;
}

export function decodeImageMetadata(data: Uint8Array): ImageMetadata {
  if (startsWith(data, [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a])) {
    const [width, height] = pngSize(data);
    return metadata(width, height, "png", pngIsAnimated(data));
  }
  if (ascii(data, 0, 2) === "BM") {
    const [width, height] = bmpSize(data);
    return metadata(width, height, "bmp", false);
  }
  const gifHeader = ascii(data, 0, 6);
  if (gifHeader === "GIF87a" || gifHeader === "GIF89a") {
    const [width, height] = gifSize(data);
    return metadata(width, height, "gif", includesAscii(data, "NETSCAPE2.0"));
  }
  if (startsWith(data, [0xff, 0xd8])) {
    const [width, height] = jpegSize(data);
    return metadata(width, height, "jpeg", false);
  }
  if (ascii(data, 0, 4) === "RIFF" && ascii(data, 8, 4) === "WEBP") {
    const [width, height, animated] = webpSize(data);
    return metadata(width, height, "webp", animated);
  }
  throw new Error("unsupported or malformed image resource");
}

function metadata(
  width: number,
  height: number,
  format: ImageMetadata["format"],
  animated: boolean,
): ImageMetadata {
  if (!Number.isInteger(width) || !Number.isInteger(height) || width <= 0 || height <= 0)
    throw new Error("image dimensions are out of range");
  return { width, height, format, animated };
}

function pngSize(data: Uint8Array): [number, number] {
  if (data.length < 24 || ascii(data, 12, 4) !== "IHDR") throw new Error("malformed PNG header");
  const view = viewOf(data);
  return [view.getUint32(16), view.getUint32(20)];
}

function pngIsAnimated(data: Uint8Array): boolean {
  const view = viewOf(data);
  let offset = 8;
  while (offset + 12 <= data.length) {
    const length = view.getUint32(offset);
    const end = offset + 12 + length;
    if (end > data.length) return false;
    const kind = ascii(data, offset + 4, 4);
    if (kind === "acTL") return true;
    if (kind === "IEND") return false;
    offset = end;
  }
  return false;
}

function bmpSize(data: Uint8Array): [number, number] {
  if (data.length < 26) throw new Error("malformed BMP header");
  const view = viewOf(data);
  const dibSize = view.getUint32(14, true);
  if (dibSize === 12) return [view.getUint16(18, true), view.getUint16(20, true)];
  if (dibSize < 40) throw new Error("unsupported BMP header");
  return [Math.abs(view.getInt32(18, true)), Math.abs(view.getInt32(22, true))];
}

function gifSize(data: Uint8Array): [number, number] {
  if (data.length < 10) throw new Error("malformed GIF header");
  const view = viewOf(data);
  return [view.getUint16(6, true), view.getUint16(8, true)];
}

function jpegSize(data: Uint8Array): [number, number] {
  let offset = 2;
  const view = viewOf(data);
  while (offset < data.length) {
    while (offset < data.length && data[offset] !== 0xff) offset += 1;
    while (offset < data.length && data[offset] === 0xff) offset += 1;
    if (offset >= data.length) break;
    const marker = data[offset++];
    if (marker === 0x01 || (marker >= 0xd0 && marker <= 0xd9)) continue;
    if (offset + 2 > data.length) break;
    const length = view.getUint16(offset);
    if (length < 2 || offset + length > data.length) break;
    const isStartOfFrame =
      marker >= 0xc0 && marker <= 0xcf && marker !== 0xc4 && marker !== 0xc8 && marker !== 0xcc;
    if (isStartOfFrame) {
      if (length < 7) break;
      return [view.getUint16(offset + 5), view.getUint16(offset + 3)];
    }
    offset += length;
  }
  throw new Error("malformed JPEG header");
}

function webpSize(data: Uint8Array): [number, number, boolean] {
  const view = viewOf(data);
  let offset = 12;
  while (offset + 8 <= data.length) {
    const kind = ascii(data, offset, 4);
    const length = view.getUint32(offset + 4, true);
    const payload = offset + 8;
    if (payload + length > data.length) break;
    if (kind === "VP8X" && length >= 10) {
      return [
        1 + uint24le(data, payload + 4),
        1 + uint24le(data, payload + 7),
        Boolean(data[payload] & 0x02),
      ];
    }
    if (kind === "VP8 " && length >= 10 && startsWith(data, [0x9d, 0x01, 0x2a], payload + 3)) {
      return [
        view.getUint16(payload + 6, true) & 0x3fff,
        view.getUint16(payload + 8, true) & 0x3fff,
        false,
      ];
    }
    if (kind === "VP8L" && length >= 5 && data[payload] === 0x2f) {
      const bits = view.getUint32(payload + 1, true);
      return [1 + (bits & 0x3fff), 1 + ((bits >>> 14) & 0x3fff), false];
    }
    offset += 8 + length + (length & 1);
  }
  throw new Error("malformed WebP header");
}

function viewOf(data: Uint8Array): DataView {
  return new DataView(data.buffer, data.byteOffset, data.byteLength);
}

function startsWith(data: Uint8Array, expected: number[], offset = 0): boolean {
  return expected.every((byte, index) => data[offset + index] === byte);
}

function ascii(data: Uint8Array, offset: number, length: number): string {
  if (offset + length > data.length) return "";
  return String.fromCharCode(...data.subarray(offset, offset + length));
}

function includesAscii(data: Uint8Array, needle: string): boolean {
  const bytes = [...needle].map((character) => character.charCodeAt(0));
  outer: for (let offset = 0; offset + bytes.length <= data.length; offset += 1) {
    for (let index = 0; index < bytes.length; index += 1)
      if (data[offset + index] !== bytes[index]) continue outer;
    return true;
  }
  return false;
}

function uint24le(data: Uint8Array, offset: number): number {
  return data[offset] | (data[offset + 1] << 8) | (data[offset + 2] << 16);
}

File: src-tauri/src/image_metadata.rs
mod bytes;

use bytes::{be_u16, be_u32, le_i32, le_u16, le_u32, uint24_le};

pub struct ImageMetadata {
    pub width: u32,
    pub height: u32,
    pub format: &'static str,
    pub animated: bool,
}

pub fn decode(data: &[u8]) -> Option<ImageMetadata> {
    if data.starts_with(&[0x89, b'P', b'N', b'G', 0x0d, 0x0a, 0x1a, 0x0a]) {
        return png(data);
    }
    if data.starts_with(b"BM") {
        return bmp(data);
    }
    if data.starts_with(b"GIF87a") || data.starts_with(b"GIF89a") {
        return gif(data);
    }
    if data.starts_with(&[0xff, 0xd8]) {
        return jpeg(data);
    }
    if data.starts_with(b"RIFF") && data.get(8..12) == Some(b"WEBP") {
        return webp(data);
    }
    None
}

fn metadata(
    width: u32,
    height: u32,
    format: &'static str,
    animated: bool,
) -> Option<ImageMetadata> {
    (width > 0 && height > 0).then_some(ImageMetadata {
        width,
        height,
        format,
        animated,
    })
}

fn png(data: &[u8]) -> Option<ImageMetadata> {
    if data.get(12..16) != Some(b"IHDR") {
        return None;
    }
    let width = be_u32(data, 16)?;
    let height = be_u32(data, 20)?;
    let mut offset = 8usize;
    let mut animated = false;
    while offset.checked_add(12)? <= data.len() {
        let length = usize::try_from(be_u32(data, offset)?).ok()?;
        let end = offset.checked_add(12)?.checked_add(length)?;
        if end > data.len() {
            break;
        }
        match data.get(offset + 4..offset + 8) {
            Some(b"acTL") => {
                animated = true;
                break;
            }
            Some(b"IEND") => break,
            _ => offset = end,
        }
    }
    metadata(width, height, "png", animated)
}

fn bmp(data: &[u8]) -> Option<ImageMetadata> {
    let dib_size = le_u32(data, 14)?;
    let (width, height) = if dib_size == 12 {
        (u32::from(le_u16(data, 18)?), u32::from(le_u16(data, 20)?))
    } else if dib_size >= 40 {
        (
            le_i32(data, 18)?.unsigned_abs(),
            le_i32(data, 22)?.unsigned_abs(),
        )
    } else {
        return None;
    };
    metadata(width, height, "bmp", false)
}

fn gif(data: &[u8]) -> Option<ImageMetadata> {
    let width = u32::from(le_u16(data, 6)?);
    let height = u32::from(le_u16(data, 8)?);
    metadata(
        width,
        height,
        "gif",
        data.windows(b"NETSCAPE2.0".len())
            .any(|window| window == b"NETSCAPE2.0"),
    )
}

fn jpeg(data: &[u8]) -> Option<ImageMetadata> {
    let mut offset = 2usize;
    while offset < data.len() {
        while data.get(offset).is_some_and(|byte| *byte != 0xff) {
            offset += 1;
        }
        while data.get(offset) == Some(&0xff) {
            offset += 1;
        }
        let marker = *data.get(offset)?;
        offset += 1;
        if marker == 0x01 || (0xd0..=0xd9).contains(&marker) {
            continue;
        }
        let length = usize::from(be_u16(data, offset)?);
        let end = offset.checked_add(length)?;
        if length < 2 || end > data.len() {
            return None;
        }
        let is_start_of_frame =
            (0xc0..=0xcf).contains(&marker) && !matches!(marker, 0xc4 | 0xc8 | 0xcc);
        if is_start_of_frame {
            if length < 7 {
                return None;
            }
            return metadata(
                u32::from(be_u16(data, offset + 5)?),
                u32::from(be_u16(data, offset + 3)?),
                "jpeg",
                false,
            );
        }
        offset = end;
    }
    None
}

fn webp(data: &[u8]) -> Option<ImageMetadata> {
    let mut offset = 12usize;
    while offset.checked_add(8)? <= data.len() {
        let kind = data.get(offset..offset + 4)?;
        let length = usize::try_from(le_u32(data, offset + 4)?).ok()?;
        let payload = offset.checked_add(8)?;
        let end = payload.checked_add(length)?;
        if kind == b"VP8X" && length >= 10 && payload.checked_add(10)? <= data.len() {
            return metadata(
                uint24_le(data, payload + 4)?.checked_add(1)?,
                uint24_le(data, payload + 7)?.checked_add(1)?,
                "webp",
                data[payload] & 0x02 != 0,
            );
        }
        if kind == b"VP8 "
            && length >= 10
            && payload.checked_add(10)? <= data.len()
            && data.get(payload + 3..payload + 6) == Some(&[0x9d, 0x01, 0x2a])
        {
            return metadata(
                u32::from(le_u16(data, payload + 6)? & 0x3fff),
                u32::from(le_u16(data, payload + 8)? & 0x3fff),
                "webp",
                false,
            );
        }
        if kind == b"VP8L"
            && length >= 5
            && payload.checked_add(5)? <= data.len()
            && data.get(payload) == Some(&0x2f)
        {
            let bits = le_u32(data, payload + 1)?;
            return metadata(
                (bits & 0x3fff) + 1,
                ((bits >> 14) & 0x3fff) + 1,
                "webp",
                false,
            );
        }
        if end > data.len() {
            return None;
        }
        offset = end.checked_add(length & 1)?;
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_png_dimensions() {
        let mut png = vec![0; 24];
        png[..8].copy_from_slice(&[0x89, b'P', b'N', b'G', 0x0d, 0x0a, 0x1a, 0x0a]);
        png[12..16].copy_from_slice(b"IHDR");
        png[16..20].copy_from_slice(&320_u32.to_be_bytes());
        png[20..24].copy_from_slice(&180_u32.to_be_bytes());
        let decoded = decode(&png).unwrap();
        assert_eq!((decoded.width, decoded.height), (320, 180));
        assert_eq!(decoded.format, "png");
        assert!(!decoded.animated);
    }

    #[test]
    fn reads_lossless_webp_dimensions_from_a_bounded_prefix() {
        let bits = (640_u32 - 1) | ((480_u32 - 1) << 14);
        let mut webp = b"RIFF\0\0\0\0WEBPVP8L".to_vec();
        webp.extend_from_slice(&(2_u32 * 1024 * 1024).to_le_bytes());
        webp.push(0x2f);
        webp.extend_from_slice(&bits.to_le_bytes());

        let decoded = decode(&webp).unwrap();
        assert_eq!((decoded.width, decoded.height), (640, 480));
        assert_eq!(decoded.format, "webp");
        assert!(!decoded.animated);
    }
}
